docker/stan-wasm-server/logic/compilation.py

```diff
@@ -38,60 +37,46 @@ def copy_compilation_outputs(*, model_dir: Path, job_dir: Path):
     for (src, dest) in todos:
         copy2(src, dest)
 
+async def compile_and_cache(*, job_id: str, model_dir: Path, tinystan_dir: Path):
+
+    # if there's a cache hit, make sure any copying is already complete,
+    # then return
+    if all((model_dir / x).exists() for x in COMPILATION_OUTPUTS):
+        wait_until_free(model_dir)
+        return
+
+    # otherwise, compile in our job-specific folder
+    await compile_stan_program(job_id=job_id, tinystan_dir=tinystan_dir)
+
+    # then, try to copy into the cache
+    with compilation_output_lock(model_dir) as exclusive:
+        # if we succeed in getting the lock, do the copy
+        if exclusive:
+            job_dir = get_compilation_job_dir(job_id)
+            for file in COMPILATION_OUTPUTS:
+                copy2(job_dir / file, model_dir / file)
+        # otherwise, just wait for the other thread's version to be
+        # copied. We wasted some time, but that's ultimately okay
+        else:
+            wait_until_free(model_dir)
```

Comment:
Might want to explicitly note that we intentionally do not copy in this case (because somebody else succeeded at compiling something semantically equivalent)